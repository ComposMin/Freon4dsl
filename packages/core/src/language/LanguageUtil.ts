import { Language } from "../storage/index";
import { PiBinaryExpression } from "./PiBinaryExpression";
import { PiElement } from "./PiElement";
import { PiExpression } from "./PiExpression";
import { PiModel } from "./PiModel";
import { PiModelUnit } from "./PiModelUnit";

export function isPiModel(element: PiElement): element is PiModel {
    return !!element && element.piIsModel && element.piIsModel();
}

export function isPiExpression(element: PiElement): element is PiExpression {
    return !!element && element.piIsExpression && element.piIsExpression();
}

export function isPiBinaryExpression(element: PiElement): element is PiBinaryExpression {
    return !!element && element.piIsExpression && element.piIsExpression() && element.piIsBinaryExpression && element.piIsBinaryExpression();
}

export function ownerOfType(element: PiElement, typename: string, exact?: boolean): PiElement {
    let parent = element.piOwnerDescriptor()?.owner;
    while (!!parent) {
        if (exact ? instanceOfExact(parent, typename) : instanceOfSub(parent, typename)) {
            return parent;
        } else {
            parent = parent.piOwnerDescriptor()?.owner;
        }
    }
    return null;
}

/**
 * Returns the modelunit that owns this `element'
 * @param element
 */
export function modelUnit(element: PiElement): PiModelUnit {
    let current = element;
    while (!!current) {
        if (current.piIsUnit()) {
            return current as PiModelUnit;
        } else {
            current = current.piOwnerDescriptor()?.owner;
        }
    }
    // No modelunit found, element is standalone
    return null;
}

export function instanceOfExact(src: PiElement, target: string): boolean {
    return src.piLanguageConcept() === target;
}

export function instanceOfSub(src: PiElement, target: string): boolean {
    return isSubConcept(src.piLanguageConcept(), target);
}

export function isExactConcept(src, target: string): boolean {
    return src === target;
}

export function isSubConcept(src, target: string): boolean {
    return src === target || Language.getInstance().classifier(target).subConceptNames.includes(src);
}

export function matchElementList(list: PiElement[], toBeMatched: Partial<PiElement>[]): boolean {
    let foundMatch: boolean = true;
    for (const theirs of toBeMatched) {
        let xx: boolean = false;
        for (const mine of list) {
            if (mine.match(theirs)) {
                xx = true;
                break;
            }
        }
        foundMatch = foundMatch && xx;
        if (!foundMatch) {
            return false;
        }
    }
    return foundMatch;
}
