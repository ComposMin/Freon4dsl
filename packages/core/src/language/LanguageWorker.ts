/**
 * Interface LanguageWorker defines the extended visitor pattern of instances of languages.
 * Every node is visited twice, once before the visit of its children, and once after this visit.
 */
import { PiElement } from "./PiElement";

export interface LanguageWorker {
    execBefore(modelelement: PiElement): boolean;
    execAfter(modelelement: PiElement): boolean;

}
