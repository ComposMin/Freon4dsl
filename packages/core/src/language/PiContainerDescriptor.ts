import { PiElement } from "./PiElement";

export interface PiContainerDescriptor {
    container: PiElement;
    propertyName: string;
    propertyIndex?: number;
}
