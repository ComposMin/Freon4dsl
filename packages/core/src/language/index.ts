export * from "./PiElement";
export * from "./PiNamedElement";
export * from "./PiExpression";
export * from "./PiBinaryExpression";
export * from "./PiContainerDescriptor";
export * from "./PiExpression";
export * from "./PiModel";
export * from "./decorators";
export * from "./LanguageUtil";
