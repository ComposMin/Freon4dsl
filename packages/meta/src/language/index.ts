export * from "./MetaGenerator";
export * from "./MetaModel";
export * from "./PiMetaTemplates";
export * from "./TestGenerator";
