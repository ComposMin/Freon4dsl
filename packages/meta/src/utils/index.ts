export * from "./Checker";
export * from "./Helpers";
export * from "./Names";
export * from "./PathProvider";
export * from "./PiParser";
export * from "./ModelHelpers";
