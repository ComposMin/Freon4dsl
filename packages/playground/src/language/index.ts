// Generated by the ProjectIt Language Generator.
export * from "./DemoAbsExpression";
export * from "./DemoAndExpression";
export * from "./DemoAttribute";
export * from "./DemoAttributeType";
export * from "./DemoBinaryExpression";
export * from "./DemoBooleanLiteralExpression";
export * from "./DemoComparisonExpression";
export * from "./DemoDivideExpression";
export * from "./DemoEntity";
export * from "./DemoEqualsExpression";
export * from "./DemoExpression";
export * from "./DemoFunction";
export * from "./DemoFunctionCallExpression";
export * from "./DemoGreaterThenExpression";
export * from "./DemoIfExpression";
export * from "./DemoLessThenExpression";
export * from "./DemoLiteralExpression";
export * from "./DemoModel";
export * from "./DemoMultiplyExpression";
export * from "./DemoNumberLiteralExpression";
export * from "./DemoOrExpression";
export * from "./DemoPlaceholderExpression";
export * from "./DemoPlusExpression";
export * from "./DemoStringLiteralExpression";
export * from "./DemoType";
export * from "./DemoVariable";
export * from "./DemoVariableRef";
export * from "./WithType";
