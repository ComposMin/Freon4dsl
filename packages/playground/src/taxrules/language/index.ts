// Generated by the ProjectIt Language Generator.
export * from "./AbsExpression";
export * from "./AllTaxRulesConcepts";
export * from "./AndExpression";
export * from "./BinaryExpression";
export * from "./BooleanLiteralExpression";
export * from "./ComparisonExpression";
export * from "./DivideExpression";
export * from "./EqualsExpression";
export * from "./Expression";
export * from "./GreaterThenExpression";
export * from "./IfExpression";
export * from "./IncomePart";
export * from "./IncomeType";
export * from "./LessThenExpression";
export * from "./LiteralExpression";
export * from "./Money";
export * from "./MoneyLiteralExp";
export * from "./MultiplyExpression";
export * from "./NumberLiteralExpression";
export * from "./OrExpression";
export * from "./PercentageExpression";
export * from "./PlaceholderExpression";
export * from "./PlusExpression";
export * from "./RevenueService";
export * from "./RuleParameter";
export * from "./StringLiteralExpression";
export * from "./TaxID";
export * from "./TaxPayer";
export * from "./TaxPayerType";
export * from "./TaxRule";
export * from "./TaxRuleCall";
export * from "./TaxRuleSet";
