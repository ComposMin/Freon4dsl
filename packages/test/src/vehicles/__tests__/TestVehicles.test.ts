import { FileHandler } from "../../utils/FileHandler";
import { VehiclesEnvironment } from "../environment/gen/VehiclesEnvironment";
import { VehicleUnit } from "../language/gen";

describe("Vehicles parser on", () => {
    const reader = VehiclesEnvironment.getInstance().reader;
    const writer = VehiclesEnvironment.getInstance().writer;
    const fileHandler = new FileHandler();

    test( " test1 ", () => {
        try {
            const input = fileHandler.stringFromFile("src/vehicles/__inputs__/test1.veh");
            const unit1: VehicleUnit = reader.readFromString(input, "VehicleUnit") as VehicleUnit;
            // console.log(writer.writeToString(unit1, 0, false));
            expect(unit1).toMatchSnapshot();
        } catch (e) {
            // console.log(e.stack);
            expect(e).toBeNaN();
        }
    });

});
