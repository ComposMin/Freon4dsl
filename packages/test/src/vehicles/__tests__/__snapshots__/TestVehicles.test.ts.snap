// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vehicles parser on  test1  1`] = `
VehicleUnit {
  "$id": "ID-23",
  "$typename": "VehicleUnit",
  "_PI_vehicles": Array [
    Bike {
      "$id": "ID-8",
      "$typename": "Bike",
      "_PI_manufacturer": Manufacturer {
        "$id": "ID-7",
        "$typename": "Manufacturer",
        "CEO_name": "",
        "container": [Circular],
        "country": "",
        "name": "Gazelle",
        "propertyIndex": undefined,
        "propertyName": "manufacturer",
      },
      "container": [Circular],
      "nrOfWheels": 5,
      "propertyIndex": 0,
      "propertyName": "vehicles",
      "typeName": "Bike",
      "tyreSize": 25,
    },
    Steamboat {
      "$id": "ID-10",
      "$typename": "Steamboat",
      "_PI_manufacturer": Manufacturer {
        "$id": "ID-9",
        "$typename": "Manufacturer",
        "CEO_name": "Peter Petersen",
        "container": [Circular],
        "country": "NL",
        "name": "Altena Yachting",
        "propertyIndex": undefined,
        "propertyName": "manufacturer",
      },
      "_PI_typeOfEngine": PiElementReference {
        "_PI_pathname": Array [
          "Steam",
        ],
        "_PI_referred": EngineType {
          "$id": "ID-1",
          "$typename": "EngineType",
          "container": null,
          "name": "Steam",
          "propertyIndex": undefined,
          "propertyName": "",
        },
        "container": [Circular],
        "propertyIndex": undefined,
        "propertyName": "typeOfEngine",
        "typeName": "EngineType",
      },
      "container": [Circular],
      "isHeritage": true,
      "maxSpeed": 30,
      "nrOfPassengers": 100,
      "nrOfWheels": 0,
      "preferedWaterWay": "Donau",
      "propertyIndex": 1,
      "propertyName": "vehicles",
      "typeName": "SteamBoat",
    },
    PetrolCar {
      "$id": "ID-12",
      "$typename": "PetrolCar",
      "_PI_manufacturer": Manufacturer {
        "$id": "ID-11",
        "$typename": "Manufacturer",
        "CEO_name": "",
        "container": [Circular],
        "country": "",
        "name": "Porsche",
        "propertyIndex": undefined,
        "propertyName": "manufacturer",
      },
      "_PI_typeOfEngine": PiElementReference {
        "_PI_pathname": Array [
          "Gasoline",
        ],
        "_PI_referred": EngineType {
          "$id": "ID-2",
          "$typename": "EngineType",
          "container": null,
          "name": "Gasoline",
          "propertyIndex": undefined,
          "propertyName": "",
        },
        "container": [Circular],
        "propertyIndex": undefined,
        "propertyName": "typeOfEngine",
        "typeName": "EngineType",
      },
      "bootContent": 20,
      "container": [Circular],
      "maxSpeed": 240,
      "nrOfWheels": 4,
      "propertyIndex": 2,
      "propertyName": "vehicles",
      "typeName": "PetrolCar",
    },
    SolarCar {
      "$id": "ID-14",
      "$typename": "SolarCar",
      "_PI_manufacturer": Manufacturer {
        "$id": "ID-13",
        "$typename": "Manufacturer",
        "CEO_name": "",
        "container": [Circular],
        "country": "",
        "name": "LightYear",
        "propertyIndex": undefined,
        "propertyName": "manufacturer",
      },
      "_PI_typeOfEngine": PiElementReference {
        "_PI_pathname": Array [
          "Electricity",
        ],
        "_PI_referred": EngineType {
          "$id": "ID-3",
          "$typename": "EngineType",
          "container": null,
          "name": "Electricity",
          "propertyIndex": undefined,
          "propertyName": "",
        },
        "container": [Circular],
        "propertyIndex": undefined,
        "propertyName": "typeOfEngine",
        "typeName": "EngineType",
      },
      "chargeCapacity": 10,
      "container": [Circular],
      "maxSpeed": 10,
      "nrOfWheels": 4,
      "propertyIndex": 3,
      "propertyName": "vehicles",
      "solarPanelBrand": "someName",
      "typeName": "SolarCar",
    },
    Scooter {
      "$id": "ID-16",
      "$typename": "Scooter",
      "_PI_manufacturer": Manufacturer {
        "$id": "ID-15",
        "$typename": "Manufacturer",
        "CEO_name": "",
        "container": [Circular],
        "country": "",
        "name": "Cushman",
        "propertyIndex": undefined,
        "propertyName": "manufacturer",
      },
      "_PI_typeOfEngine": PiElementReference {
        "_PI_pathname": Array [
          "Electricity",
        ],
        "_PI_referred": EngineType {
          "$id": "ID-3",
          "$typename": "EngineType",
          "container": null,
          "name": "Electricity",
          "propertyIndex": undefined,
          "propertyName": "",
        },
        "container": [Circular],
        "propertyIndex": undefined,
        "propertyName": "typeOfEngine",
        "typeName": "EngineType",
      },
      "chargeCapacity": 10,
      "color": "someName",
      "container": [Circular],
      "maxSpeed": 10,
      "nrOfWheels": 2,
      "propertyIndex": 4,
      "propertyName": "vehicles",
      "typeName": "Scooter",
    },
    Segway {
      "$id": "ID-18",
      "$typename": "Segway",
      "_PI_manufacturer": Manufacturer {
        "$id": "ID-17",
        "$typename": "Manufacturer",
        "CEO_name": "",
        "container": [Circular],
        "country": "",
        "name": "Segway",
        "propertyIndex": undefined,
        "propertyName": "manufacturer",
      },
      "_PI_typeOfEngine": PiElementReference {
        "_PI_pathname": Array [
          "Electricity",
        ],
        "_PI_referred": EngineType {
          "$id": "ID-3",
          "$typename": "EngineType",
          "container": null,
          "name": "Electricity",
          "propertyIndex": undefined,
          "propertyName": "",
        },
        "container": [Circular],
        "propertyIndex": undefined,
        "propertyName": "typeOfEngine",
        "typeName": "EngineType",
      },
      "chargeCapacity": 10,
      "container": [Circular],
      "isAllowedOnRoad": true,
      "maxSpeed": 10,
      "nrOfWheels": 2,
      "propertyIndex": 5,
      "propertyName": "vehicles",
      "typeName": "Segway",
    },
    Scateboard {
      "$id": "ID-20",
      "$typename": "Scateboard",
      "_PI_manufacturer": Manufacturer {
        "$id": "ID-19",
        "$typename": "Manufacturer",
        "CEO_name": "",
        "container": [Circular],
        "country": "",
        "name": "Olympics",
        "propertyIndex": undefined,
        "propertyName": "manufacturer",
      },
      "container": [Circular],
      "nrOfWheels": 8,
      "propertyIndex": 6,
      "propertyName": "vehicles",
      "size": 10,
      "typeName": "Scateboard",
    },
    KickBike {
      "$id": "ID-22",
      "$typename": "KickBike",
      "_PI_manufacturer": Manufacturer {
        "$id": "ID-21",
        "$typename": "Manufacturer",
        "CEO_name": "",
        "container": [Circular],
        "country": "",
        "name": "Kicks",
        "propertyIndex": undefined,
        "propertyName": "manufacturer",
      },
      "container": [Circular],
      "foldable": false,
      "nrOfWheels": 2,
      "propertyIndex": 7,
      "propertyName": "vehicles",
      "typeName": "KickBike",
    },
  ],
  "container": null,
  "name": "unitTest1",
  "propertyIndex": undefined,
  "propertyName": "",
}
`;
